File: 3sem/4lab/Banks/Transactions/DepositTransaction.cs
using Banks.BankAccounts;
using Banks.Exceptions;
using Banks.Models;

namespace Banks.Transactions;

public class DepositTransaction : ITransaction
{
    public DepositTransaction(decimal sum, IBankAccount account)
    {
        ArgumentNullException.ThrowIfNull(account);

        Id = Guid.NewGuid();
        Status = TransactionStatus.Created;

        Sum = sum;
        Account = account;
    }

    public Guid Id { get; }
    public decimal Sum { get; }
    public IBankAccount Account { get; }
    public TransactionStatus Status { get; private set; }

    public void Execute()
    {
        if (Status is not TransactionStatus.Created)
            throw InvalidTransactionOperationException.OnIncorrectExecuting();

        if (!Account.CanDeposit(Sum))
            throw InvalidTransactionOperationException.OnInvalidTransaction();

        Account.Deposit(Sum);

        Status = TransactionStatus.Executed;
    }

    public void Cancel()
    {
        if (Status is not TransactionStatus.Executed)
            throw InvalidTransactionOperationException.OnIncorrectCanceling();

        if (!Account.CanWithdraw(Sum))
            throw InvalidTransactionOperationException.OnInvalidTransaction();

        Account.Withdraw(Sum);

        Status = TransactionStatus.Canceled;
    }
}

File: 4sem/1lab/src/main/java/banksRemake/domain/transactions/DepositTransaction.java
package banksRemake.domain.transactions;

import banksRemake.domain.bankAccounts.BankAccount;
import banksRemake.domain.exceptions.InvalidBankAccountOperationException;
import banksRemake.domain.exceptions.InvalidTransactionOperationException;
import banksRemake.domain.models.TransactionStatus;

import java.util.UUID;

/**
 * The operation of crediting money to the account.
 */
public class DepositTransaction implements Transaction {
    private final UUID id;
    private final double sum;
    private final BankAccount account;
    private TransactionStatus status;

    public DepositTransaction(double sum, BankAccount account) {
        this.status = TransactionStatus.Created;

        this.sum = sum;
        this.account = account;

        this.id = UUID.randomUUID();
    }

    @Override
    public UUID getId() {
        return id;
    }

    @Override
    public double getSum() {
        return sum;
    }

    @Override
    public TransactionStatus getStatus() {
        return status;
    }

    @Override
    public void execute() throws InvalidTransactionOperationException, InvalidBankAccountOperationException {
        if (status != TransactionStatus.Created) {
            throw InvalidTransactionOperationException.onIncorrectExecuting();
        }

        if (!account.canDeposit(sum)) {
            throw InvalidTransactionOperationException.onInvalidTransaction();
        }

        account.deposit(sum);

        status = TransactionStatus.Executed;
    }

    @Override
    public void cancel() throws InvalidTransactionOperationException, InvalidBankAccountOperationException {
        if (status != TransactionStatus.Executed){
            throw InvalidTransactionOperationException.onIncorrectCanceling();
        }

        if (!account.canWithdraw(sum)) {
            throw InvalidTransactionOperationException.onInvalidTransaction();
        }

        account.withdraw(sum);

        status = TransactionStatus.Canceled;
    }
}
